lib/virtio_mmio/virtio_mmio.c

```diff
@@ -0,0 +1,416 @@
+extern void virtqueue_notify(struct virtqueue *vq);
+extern void virtio_reset_vd(struct virtio_device *vdev);
+extern int virtqueue_empty(struct virtqueue *vq);
+extern int virtqueue_full(struct virtqueue *vq);
+
+static inline void virtio_mmio_write32(struct virtio_device *vdev, int offset, uint32_t value)
+{
+	struct virtio_mmio_device *vmdev = (struct virtio_mmio_device *)vdev->priv;
+
+	metal_io_write32(vmdev->cfg_io, offset, value);
+}
+
+static inline uint32_t virtio_mmio_read32(struct virtio_device *vdev, int offset)
+{
+	struct virtio_mmio_device *vmdev = (struct virtio_mmio_device *)vdev->priv;
+
+	return metal_io_read32(vmdev->cfg_io, offset);
+}
+
+static inline uint8_t virtio_mmio_read8(struct virtio_device *vdev, int offset)
+{
+	struct virtio_mmio_device *vmdev = (struct virtio_mmio_device *)vdev->priv;
+
+	return metal_io_read8(vmdev->cfg_io, offset);
+}
+
+static void virtio_mmio_set_status(struct virtio_device *vdev, uint8_t status)
+{
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_STATUS, status);
+}
+
+static uint8_t virtio_mmio_get_status(struct virtio_device *vdev)
+{
+	return virtio_mmio_read32(vdev, VIRTIO_MMIO_STATUS);
+}
+
+static void virtio_mmio_write_config(struct virtio_device *vdev,
+				     uint32_t offset, void *dst, int length)
+{
+	(void)(vdev);
+	(void)(offset);
+	(void)(dst);
+	(void)length;
+
+	metal_log(METAL_LOG_WARNING, "%s not supported\n", __func__);
+}
+
+static void virtio_mmio_read_config(struct virtio_device *vdev,
+				    uint32_t offset, void *dst, int length)
+{
+	int i;
+	uint8_t *d = dst;
+	(void)(offset);
+
+	for (i = 0; i < length; i++)
+		d[i] = virtio_mmio_read8(vdev, VIRTIO_MMIO_CONFIG + i);
+}
+
+static uint32_t _virtio_mmio_get_features(struct virtio_device *vdev, int idx)
+{
+	uint32_t hfeatures;
+
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_DEVICE_FEATURES_SEL, idx);
+	hfeatures = virtio_mmio_read32(vdev, VIRTIO_MMIO_DEVICE_FEATURES);
+	return hfeatures & vdev->features;
+}
+
+static uint32_t virtio_mmio_get_features(struct virtio_device *vdev)
+{
+	return _virtio_mmio_get_features(vdev, 0);
+}
+
+/* This is more like negotiate_features */
+static void _virtio_mmio_set_features(struct virtio_device *vdev,
+				      uint32_t features, int idx)
+{
+	uint32_t hfeatures;
+
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_DRIVER_FEATURES_SEL, idx);
+	hfeatures = virtio_mmio_read32(vdev, VIRTIO_MMIO_DEVICE_FEATURES);
+	features &= hfeatures;
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_DRIVER_FEATURES, features);
+	vdev->features = features;
+}
+
+static void virtio_mmio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	_virtio_mmio_set_features(vdev, features, 0);
+}
+
+static void virtio_mmio_reset_device(struct virtio_device *vdev)
+{
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_STATUS, 0);
+}
+
+static void virtio_mmio_notify(struct virtqueue *vq)
+{
+	virtio_mmio_write32(vq->vq_dev, VIRTIO_MMIO_QUEUE_NOTIFY, vq->vq_queue_index);
```

Comment:
Add a comment here specifying that the VIRTIO_F_NOTIFICATION_DATA feature is not supported.
could be nice to return an error also in _virtio_mmio_set_features